Update Spot URDF Xacro and rviz launch parameter

abaker-bdai opened the pull request:
This PR address [PN-31](https://theaiinstitute.atlassian.net/browse/PN-31) and [PN-67](https://theaiinstitute.atlassian.net/browse/PN-67)

This PR provides the following updates to the `spot_description` and `spot_driver` packages:
- Converted main spot and spot arm urdfs into xacro macros in order to add parameters.
- added parameters to include arm in urdf, define prefixes for joint/link names, optionally specify a parent frame.
- renamed `spot.urdf.xacro` to `spot_macro.xacro`
- renamed `spot_with_arm.urdf.xacro` to `spot.urdf.xacro` and refactored to use `spot_macro.xacro`
- updated launch files to use `spot_macro.xacro` 
- removed namespaces launch file as it is no longer needed.

Testing:
You can test the updates by launch the rviz with the robot model by running the following command:
``` bash
ros2 launch spot_description description.launch.py arm:=true prefix:=test_
```
If you wish to test using the spot driver, run the following command:
``` bash
ros2 launch spot_driver spot_driver.launch.py arm:=true prefix:=test_
```


[PN-31]: https://theaiinstitute.atlassian.net/browse/PN-31?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PN-67]: https://theaiinstitute.atlassian.net/browse/PN-67?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bhung-bdai commented on spot_description/launch/description.launch.py:
```diff
@@ -2,19 +2,45 @@
 
```
I don't know if I'm setting something incorrectly, but I am getting this error when I try to run the Rviz with the URDF:
![image](https://github.com/bdaiinstitute/spot_ros2/assets/137220849/fe6c2a0b-a89e-4fe8-b200-418dcdaa47ba)

abaker-bdai replied:
I forgot to set a default for a parameter, thank you for catching this!

bhung-bdai commented on spot_driver/launch/spot_driver.launch.py:
```diff
@@ -30,23 +53,41 @@ def launch_robot_state_publisher(context: LaunchContext, has_arm: LaunchConfigur
+
+    launch_args.append(
+        DeclareLaunchArgument(
+            "tf_prefix", default_value='""', description="apply namespace prefix to robot links and joints"
```
Launching `spot_ros2`  appears to be fine, but adding the `prefix:=<name>` based on the initial comments doesn't appear to change the ROS2 topic names. It also doesn't look like it's a launch argument?

The `tf_prefix` only changes the names of the segments. Is this also meant to change the topic names, like how launching with a namespace did?

abaker-bdai replied:
It does not apply the prefix to the topic name, only to the joints and links defined in the URDF. Would you prefer that they apply the namespace to the topic name as well? or have both a namespace for the topic and prefixes for each frame?

bhung-bdai replied:
I would double check this as a last step with @amessing-bdai and @jbarry-bdai to make sure the prefix behaviour is correct. Otherwise, Rviz and launching the driver work on my end

bhung-bdai commented on spot_description/urdf/spot_arm_macro.urdf:
```diff
@@ -0,0 +1,223 @@
+        <inertia ixx="0.009014303010689815" ixy="0.0" ixz="0.0" iyy="0.005835102417829581" iyz="0.0" izz="0.00778747846359954"/>
+      </inertial>
+    </link>
+    <joint name="${tf_prefix}arm_sh0" type="revolute">
```
What do you think about adding a `/` after ${tf_prefix}? Should the user be responsible for that?

bhung-bdai reviewed (changes requested):
Some unexpected behavior when I run it. Am I using it incorrectly?

bhung-bdai commented:
If the remote testing isn't set up yet, feel free to let me know when you want changes tested on HW.

abaker-bdai commented:
This is ready for another round of review. I realized some of my changes were unnecessary and removed them. I also added a joint_state_publisher node to the spot_driver launch file so that the robot model in rviz updates properly.

bhung-bdai reviewed (approved):
I'll approve the changes for now since it's working as intended, but please address the requested comments to see if there needs to be a change.

bhung-bdai commented on spot_driver/launch/spot_driver.launch.py:
```diff
@@ -2,51 +2,134 @@
-    ld.add_action(spot_driver_node)
 
-    ld.add_action(OpaqueFunction(function=launch_robot_state_publisher, args=[has_arm, ld]))
+    launch_args.append(DeclareLaunchArgument("spot_name", default_value="", description="Name of Spot"))
```
When I set `tf_prefix:=<spot_name>/` and `spot_name:=<spot_name>` and launch with rviz, this happens:
![image](https://github.com/bdaiinstitute/spot_ros2/assets/137220849/cda87cbd-f990-441f-991b-d59192053c7c)
It appears that several of the links are improperly named:
![image](https://github.com/bdaiinstitute/spot_ros2/assets/137220849/1e7edbfe-8543-4ebb-bc58-2bc3b31a01b7)
I would hazard a guess that somewhere, it's not updating the link names correctly to reflect the changes to the `tf_prefix`.

abaker-bdai replied:
Yeah, I realize I made an incorrect assumption here and that the joint states, and consequently the joint names, are published via `spot_ros2` and so the tf_prefix will break the mapping and should be removed

bhung-bdai commented on spot_driver/launch/spot_driver.launch.py:
```diff
@@ -2,51 +2,134 @@
+        # Add robot models for each robot
+        for display in config["Visualization Manager"]["Displays"]:
+            if "RobotModel" in display["Class"]:
+                display["Description Topic"]["Value"] = f"/{robot_name}/robot_description"
```
when `{robot_name}` is not provided, it errors looking for the model because the topic looks for `/robot_description`:
![image](https://github.com/bdaiinstitute/spot_ros2/assets/137220849/a94f6546-97a2-4590-a55b-0973c41d4f9a)

bhung-bdai reviewed (changes requested):
The namespace has some bugs. Hopefully fixing won't be too much of an issue. I've provided some screenshots with the bugs being shown.

bhung-bdai reviewed (changes requested):
There's a bug where the `spot_name` ROS param isn't applied to the arm links in the URDF, but are applied in the tf tree: 
![image](https://github.com/bdaiinstitute/spot_ros2/assets/137220849/cf629d83-4dab-473c-95ba-9134d0d36041)
![image](https://github.com/bdaiinstitute/spot_ros2/assets/137220849/82414687-f3fe-469c-bb5d-7bd4317a1429)

The other functions look like they are working ok.

bhung-bdai reviewed (approved):
LGTM. (Namespace, no namespace) with (arm, no arm) both work in visualization.

abaker-bdai commented:
I'll submit a PR to update the bdai repo
